File: code/socket2/python/echo_EPLTserv.py
import socket
import select
import sys

BUF_SIZE = 4
EPOLL_SIZE = 50


def main():
    if len(sys.argv) != 2:
        print(f"Usage : {sys.argv[0]} <port>")
        sys.exit(1)

    serv_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    serv_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    serv_sock.bind(("", int(sys.argv[1])))
    serv_sock.listen(5)

    epfd = select.epoll()
    epfd.register(serv_sock.fileno(), select.EPOLLIN)

    # fd -> socket 对象 的映射，方便根据 fd 反查 socket
    fd_to_sock = {serv_sock.fileno(): serv_sock}

    try:
        while True:
            events = epfd.poll(EPOLL_SIZE, -1)
            if not events:
                continue

            print("return epoll_wait")
            for fd, event in events:
                if fd == serv_sock.fileno():
                    # 有新客户端连接
                    clnt_sock, clnt_adr = serv_sock.accept()
                    epfd.register(clnt_sock.fileno(), select.EPOLLIN)
                    fd_to_sock[clnt_sock.fileno()] = clnt_sock
                    print(f"connected client: {clnt_sock.fileno()}")
                else:
                    clnt_sock = fd_to_sock.get(fd)
                    if clnt_sock is None:
                        continue

                    try:
                        data = clnt_sock.recv(BUF_SIZE)
                    except ConnectionResetError:
                        data = b""

                    if not data:
                        # 客户端关闭连接
                        epfd.unregister(fd)
                        clnt_sock.close()
                        del fd_to_sock[fd]
                        print(f"closed client: {fd}")
                    else:
                        # echo 回去
                        clnt_sock.sendall(data)
    except KeyboardInterrupt:
        pass
    finally:
        serv_sock.close()
        epfd.close()


if __name__ == "__main__":
    main()

File: code/socket2/cxx/echo_EPLTserv.cpp
#include "socket.hpp"

#include <cstdlib>
#include <iostream>
#include <string>
#include <unordered_map>

using namespace unpsock;

int main(int argc, char* argv[])
{
    if (argc != 2) {
        std::cerr << "Usage : " << argv[0] << " <port>\n";
		exit(1);
    }

    // ---- 监听 socket ----
    Socket serv_sock(Family::INET, AddrType::STREAM);
    serv_sock.set_reuse_addr(true);

    auto serv_addr = Address::any(std::atoi(argv[1]), Family::INET);
    serv_sock.bind(serv_addr);
    serv_sock.listen(5);

    // ---- Selector ----
    EpollSelector sel;
    sel.init();
    sel.add(serv_sock, Event::Read);

    std::unordered_map<int, Socket> clients;

    const int BUF_SIZE = 4;
    while (true) {
        // 永久阻塞（原程序 epoll_wait(..., -1)）
        auto ready = sel.wait();   // timeout = nullopt
        std::cout << "return epoll_wait" << std::endl;

        for (const auto& ev : ready) {
            int fd = ev.fd;

            // 异常/挂断处理
            if (has_event(ev.events, Event::Error) ||
                    has_event(ev.events, Event::Hangup)) {
                if (fd != serv_sock.fileno()) {
                    sel.remove(fd);
                    clients.erase(fd);
                    std::cout << "closed client: " << fd << std::endl;
                }
                continue;
            }

            if (!has_event(ev.events, Event::Read)) continue;

            if (fd == serv_sock.fileno()) {
                // ---- 新连接 ----
                auto [conn, peer] = serv_sock.accept();
                int conn_fd = conn.fileno();
                sel.add(conn, Event::Read);
                clients.emplace(conn_fd, std::move(conn));
                std::cout << "connected client: " << conn_fd << std::endl;
            } else {
                // ---- 客户端数据 ----
                auto it = clients.find(fd);
                if (it == clients.end()) continue;

                Socket& conn = it->second;
                auto data = conn.recv(BUF_SIZE);

                if (data.empty()) {
                    // 对端关闭
                    sel.remove(fd);
                    clients.erase(it);
                    std::cout << "closed client: " << fd << std::endl;
                } else {
                    conn.sendall(data);     // echo
                }
            }
        }
    }

    return 0;
}
